### PR TITLE
Make Matcher more testable

### PR DESCRIPTION
**Pull request by troydai**
Addresses issue #60 .

@DamianEdwards

### CODE REVIEW

File: src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs [hunk 1]
```diff
@@ -10,22 +10,22 @@ namespace Microsoft.Framework.FileSystemGlobbing
 {
     public class Matcher
     {
-        private IList<IPattern> _includePatterns = new List<IPattern>();
-        private IList<IPattern> _excludePatterns = new List<IPattern>();
+        protected IList<IPattern> _includePatterns = new List<IPattern>();
+        protected IList<IPattern> _excludePatterns = new List<IPattern>();
```
Comment (DamianEdwards) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: Should we just go ahead and make these properties? Seems strange to have protected fields with _ prefix naming.

File: src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs [hunk 2]
```diff
@@ -10,22 +10,22 @@ namespace Microsoft.Framework.FileSystemGlobbing
 {
     public class Matcher
     {
-        private IList<IPattern> _includePatterns = new List<IPattern>();
-        private IList<IPattern> _excludePatterns = new List<IPattern>();
+        protected IList<IPattern> _includePatterns { get; set; } = new List<IPattern>();
```
Comment (Praburaj) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: variable names to be like properties? `IncludePatterns` & `ExcludePatterns`?

Comment (troydai) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: Yeah. hasn't changed it.

**Comment by DamianEdwards** [comment]
:shipit:

File: src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs [hunk 3]
```diff
@@ -10,24 +10,24 @@ namespace Microsoft.Framework.FileSystemGlobbing
 {
     public class Matcher
     {
-        private IList<IPattern> _includePatterns = new List<IPattern>();
-        private IList<IPattern> _excludePatterns = new List<IPattern>();
+        protected IList<IPattern> IncludePatterns { get; set; } = new List<IPattern>();
+        protected IList<IPattern> ExcludePatterns { get; set; } = new List<IPattern>();
```
Comment (Eilon) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: Protected is not a great pattern. Why not public?

Comment (troydai) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: The collection is not essential to external consumption. If it is public, then we need to guard the situations in which a invalid pattern is added to the pattern collection.

Comment (Eilon) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: How do you guard against a derived class from adding invalid patterns? Changing a class's design *just* for a unit test is a slippery slope.

Comment (DamianEdwards) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: I wondered whether the properties shouldn't just be IEnumerable to enable inspection, and leave the Add/Remove as distinct APIs

-----Original Message-----

Comment (DamianEdwards) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: Or maybe the fields don't need to be exposed at all and we just rely on the methods being mockable

Comment (Eilon) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: :+1: to what @DamianEdwards said.

Comment (troydai) [inline src/Microsoft.Framework.FileSystemGlobbing/Matcher.cs]: so, private fields. virtual methods.

**Comment by Eilon** [comment]
Looks cleaner to me!

**Comment by troydai** [comment]
Assume it's a :shipit: ?

**Comment by Eilon** [comment]
:shipit: !

**Pull request merged**